packages/cortex-debug/src/node/debug-cortex.ts

```diff
@@ -0,0 +1,212 @@
+/********************************************************************************
+ * Copyright (C) 2018 Arm and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { DebugConfiguration } from '@theia/debug/lib/common/debug-common';
+import { AbstractVSCodeDebugAdapterContribution } from '@theia/debug/lib/node/vscode/vscode-debug-adapter-contribution';
+import { join } from 'path';
+import { existsSync } from 'fs';
+
+const OPENOCD_VALID_RTOS: string[] = ['eCos', 'ThreadX', 'FreeRTOS', 'ChibiOS', 'embKernel', 'mqx', 'uCOS-III'];
+const JLINK_VALID_RTOS: string[] = ['FreeRTOS', 'embOS'];
+
+@injectable()
+export class CortexDebugAdapterContribution extends AbstractVSCodeDebugAdapterContribution {
+
+    constructor() {
+        super(
+            'cortex-debug',
+            join(__dirname, '../../adapters/cortex-debug/extension')
+        );
+    }
+
+    public async resolveDebugConfiguration(config: DebugConfiguration, workspaceFolderUri?: string): Promise<DebugConfiguration | undefined> {
+
+        if (!config.request) {
+            config.request = 'launch';
+        }
+
+        /**
+         * Required settings
+         */
+        if (!config.executable) {
+            throw new Error('Executable is not provided.');
+        }
+
+        if (config.cwd) {
+            config.executable = join(config.cwd, config.executable);
```

Comment:
What if config.executable is already an absolute path?

Reply:
Good point :)

Reply:
Fixed

Reply:
Also, I didn't know how to word this, but I don't think this is very intuitive.  At least I wouldn't expect the executable to be appended to `cwd`.  In my mind, `cwd` would be the directory on the remote board (assuming the remote board has a filesystem) to use as the current working directory of the newly created process.  Whereas `executable` is a path on the development system.  So appending those two wouldn't make sense.

Reply:
That's not how embedded development/debug works. Devices usually run a single binary and there isn't a remote filesystem (think firmware). In this adapter, `cwd` can be set which is the source root of the project you are debugging and executable is the path to the elf file (binary + symbols) which is uploaded to the device for debugging.

Reply:
In this particular case, there is no remote filesystem.  But when we'll do remote debugging of Linux systems, there will be.  In that case, I guess we'll want `cwd` to denote the cwd on the remote filesystem.  I would find it confusing to have different meanings of `cwd` in different launch types...

Reply:
I was under the impression the settings and what they mean are the decision of the debug adapter developer. Would we be expected to potentially rewrite any adapter we bring into Theia to align with our guidelines?

Reply:
Ah ok, I see that this is then used by the debug adapter itself:

https://marcelball.ca/projects/cortex-debug/cortex-debug-launch-configurations/

I thought it was only used by your code.  Then we have to live with this definition, I am fine with what you did given the situation.